### PR TITLE
Adding vmssPublicIp related features [examples fail due to https://github.com/Azure/oav/issues/115]

### PR DESCRIPTION
**Pull request by EvgenyAgafonchikov**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.

Spec extractions:
 - List all publicipaddress on a vmscaleset level 

    - Sample URI: https://management.azure.com/subscriptions/SubId/resourceGroups/VMSSIpTest/providers/Microsoft.Compute/virtualMachineScaleSets/vmss1/publicipaddresses?api-version=2017-03-30 

    - Return type is a LIST OF PUBLICIPADDRESSES 

 - List publicipaddress
    - https://management.azure.com/subscriptions/SubId/resourceGroups/VMSSIpTest/providers/Microsoft.Compute/virtualMachineScaleSets/vmss1/virtualMachines/0/networkInterfaces/nic1/ipconfigurations/ip1/publicipaddresses?api-version=2017-03-30 

    - Return type is a LIST OF PUBLICIPADDRESSES 

 - Get an individual publicipaddress
    - https://management.azure.com/subscriptions/SubId/resourceGroups/VMSSIpTest/providers/Microsoft.Compute/virtualMachineScaleSets/vmss1/virtualMachines/0/networkInterfaces/nic1/ipconfigurations/ip1/publicipaddresses/publicip1?api-version=2017-03-30 
    - return type is a single publicipaddress 

- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [?] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by EvgenyAgafonchikov** [comment]
@vishrutshah @veronicagg, could you please clarify this step: https://github.com/Azure/autorest/wiki/linting
I have errors even for existing specs when running `autorest -CodeGenerator None -Input arm-network\2017-03-01\swagger\<specFile>.json`. Was something changed there?

**Comment by vishrutshah** [comment]
Hi @EvgenyAgafonchikov, 

It is possible that swagger you are updating is not clean from linter errors. If you are making changes into existing ones, please looks for the errors before your change and after you changes to identify whether your changes are introducing any new errors or not.

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 1]
```diff
@@ -0,0 +1,252 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "NetworkManagementClient",
+    "description": "The Microsoft Azure Network management API provides a RESTful set of web services that interact with Microsoft Azure Networks service to manage your network resources. The API has entities that capture the relationship between an end user and the Microsoft Azure Networks service.",
+    "version": "2017-03-01"
```
Comment (DeepakRajendranMsft) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: can you change this to 2017-03-30 and see what happens please? this was suggested by sdk team as workaround but not sure if it is still a problem. Resources.azure.com have a problem with this

**Review by DeepakRajendranMsft — APPROVED** [review]

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json (hunk 1 above)
Comment (DeepakRajendranMsft) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: see what happens => does SDK generation work

**Comment by EvgenyAgafonchikov** [comment]
Hi @vishrutshah,

I'm adding new specs file and have 3 kinds of errors there. What of them could be ignored now and what have to be fixed?

1. The same for existing files, say virtualNetworks.json, and the file I'm adding:
```
ERROR: Operations API must be implemented for '/providers/Microsoft.Compute/operations'.
        Path: arm-network\2017-03-01\swagger\vmssPublicIpAddress.json#/paths
```

2. Errors related to the external references (e.g. SecurityRule, Subnet, etc.) that are not described in the file I'm adding.

3. Examples related: `ERROR: Please provide x-ms-examples describing minimum/maximum property set`. Currently I cannot get examples to work because of similar failure: https://github.com/Azure/oav/issues/115

**Comment by vishrutshah** [comment]
For 1, I see that you are putting `Microsoft.Compute` provider in the `arm-network`. Could you please help me understand why? and then i'd check and see whether `arm-compute` has `/providers/Microsoft.Compute/operations` API. If so then you should be fine ignoring. 

For 2, I'd go case by case understanding whether those errors makes sense for your operation or not. Example: [RequiredPropertiesMissingInResourceModel](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/238667689#L1342), you should check whether this is a tracked / proxy resource? And if so, Does it need to have `id`, `name` & `type` ? 

For 3, You can add examples but put the link to the bug while we fix it. 

Thanks,

**Comment by DeepakRajendranMsft** [comment]
@vishrutshah, 
for 1, please talk to Sameer and Amar, they have more context on how VMSS networking resource is modelled. Also look at https://github.com/Azure/azure-rest-api-specs/blob/master/arm-network/2017-03-01/swagger/vmssNetworkInterface.json. we are modelling publicipaddress the same way
The apis defined here are pure GET apis

for 2, They should be tracked and fixed separately and not try to be fixed here. Lets have a separate discussion this. We have already communicated with your team on required properties (please check with Samer)

For 3, these are pure GET apis, how should examples looks like? can you give us an existing sample please?

Thanks!

**Comment by salameer** [comment]
Hi Vishrut,

For Number 1, this Model actually belongs to compute but to remove cross dependency that's why it's here too.

for number 2 let's chat on the side as Deepak suggested.

@vishrutshah and @jianghaolu please follw up :)

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachineScaleSets/{virtualMachineScaleSetName}/publicipaddresses": {
+      "get": {
+        "operationId": "PublicIPAddresses_ListVirtualMachineScaleSetPublicIPAddresses",
+        "description": "Gets information about all public IP addresses on a virtual machine scale set level.",
```
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Please reference your added example like this: https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/x-ms-examples.md#structure-of-the-extension

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachineScaleSets/{virtualMachineScaleSetName}/virtualMachines/{virtualmachineIndex}/networkInterfaces/{networkInterfaceName}/ipconfigurations/{ipConfigurationName}/publicipaddresses" : {
+      "get": {
+        "operationId": "PublicIPAddresses_ListVirtualMachineScaleSetVMPublicIPAddresses",
+        "description": "Gets information about all public IP addresses in a virtual machine IP configuration in a virtual machine scale set.",
```
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Please reference your added example like this: https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/x-ms-examples.md#structure-of-the-extension

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachineScaleSets/{virtualMachineScaleSetName}/virtualMachines/{virtualmachineIndex}/networkInterfaces/{networkInterfaceName}/ipconfigurations/{ipConfigurationName}/publicipaddresses/{publicIpAddressName}": {
+      "get": {
+        "operationId": "PublicIPAddresses_GetVirtualMachineScaleSetPublicIPAddress",
+        "description": "Get the specified public IP address in a virtual machine scale set.",
```
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Please reference your added example like this: https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/x-ms-examples.md#structure-of-the-extension

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachineScaleSets/{virtualMachineScaleSetName}/publicipaddresses": {
+      "get": {
+        "operationId": "PublicIPAddresses_ListVirtualMachineScaleSetPublicIPAddresses",
```
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Consider renaming to `PublicIPAddresses_ListByVirtualMachineScaleSet`

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachineScaleSets/{virtualMachineScaleSetName}/virtualMachines/{virtualmachineIndex}/networkInterfaces/{networkInterfaceName}/ipconfigurations/{ipConfigurationName}/publicipaddresses" : {
+      "get": {
+        "operationId": "PublicIPAddresses_ListVirtualMachineScaleSetVMPublicIPAddresses",
```
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Consider renaming to `PublicIPAddresses_ListByVirtualMachineScaleSetVM`

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachineScaleSets/{virtualMachineScaleSetName}/virtualMachines/{virtualmachineIndex}/networkInterfaces/{networkInterfaceName}/ipconfigurations/{ipConfigurationName}/publicipaddresses/{publicIpAddressName}": {
+      "get": {
+        "operationId": "PublicIPAddresses_GetVirtualMachineScaleSetPublicIPAddress",
```
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Consider renaming to `PublicIPAddresses_GetByVirtualMachineScaleSet`

**Review by jianghaolu — CHANGES_REQUESTED** [review]

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json (hunk 5 above)
Comment (DeepakRajendranMsft) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: we have cannot be different to vmssNetworkInterfaces.json. so lets keep it as it is

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json (hunk 6 above)
Comment (DeepakRajendranMsft) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: we have cannot be different to vmssNetworkInterfaces.json. so lets keep it as it is

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json (hunk 7 above)
Comment (DeepakRajendranMsft) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: we have cannot be different to vmssNetworkInterfaces.json. so lets keep it as it is

**Comment by DeepakRajendranMsft** [comment]
@EvgenyAgafonchikov  please address @jianghaolu's comments on the examples.

File: arm-network/2017-03-01/swagger/vmssPublicIpAddress.json (hunk 5 above)
Comment (jianghaolu) [inline arm-network/2017-03-01/swagger/vmssPublicIpAddress.json]: Make sense. I didn't know this is migrated from an older spec.

**Comment by EvgenyAgafonchikov** [comment]
Example refs are added. @jianghaolu please have another look

**Comment by jianghaolu** [comment]
Thanks @EvgenyAgafonchikov !

One more thing I should have caught earlier (Sorry!) is that I think all these files should be in `arm-network/2017-03-30` instead of `arm-network/2017-03-01`?

**Comment by DeepakRajendranMsft** [comment]
@jianghaolu @EvgenyAgafonchikov these should still be in 2017-03-01. there is no 2017-03-30 network api version. i believe these are being handled correctly in the new repo model.

**Review by jianghaolu — APPROVED** [review]

**Pull request merged**